### PR TITLE
Update confidence histograms for better readability

### LINKED ISSUE #7257: Make confidence histograms easier to read
**Description of Problem**:

`rasa test` currently produces a histogram of confidence values for both correct and incorrect predictions. 

image

IMO something like this would be easier to read, e.g.:

image

**Definition of Done**:
- [ ] Tests are passing
- [ ] Example images are updated in the docs
- [ ] Feature mentioned in the changlog

### PR DESCRIPTION
**Pull request by lymburn**
**Proposed changes**:
Closes #7257 
- Add method to plot histogram with separate correct and incorrect prediction confidences for easier differentiation.
- Performed manual testing on Moodbot.

Image of new histogram in documentation below:   
image

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @koernerfelicia will take a look at it as soon as possible ✨

**Comment by UTkzhang** [comment]
Thanks @koernerfelicia

File: rasa/utils/plotting.py [hunk 1]
```diff
@@ -116,15 +116,15 @@ def plot_confusion_matrix(
         fig.savefig(output_file, bbox_inches="tight")
 
 
-def plot_histogram(
+def plot_comparative_histogram(
```
Comment (koernerfelicia) [inline rasa/utils/plotting.py:119]: Please preserve the original method name (`plot_histogram`) for backwards compatibility

File: rasa/nlu/test.py [hunk 2]
```diff
@@ -330,7 +330,7 @@ def plot_attribute_confidences(
         if getattr(r, target_key) != getattr(r, prediction_key)
     ]
 
-    plot_utils.plot_histogram([pos_hist, neg_hist], title, hist_filename)
+    plot_utils.plot_comparative_histogram([pos_hist, neg_hist], title, hist_filename)
```
Comment (koernerfelicia) [inline rasa/nlu/test.py:333]: Please preserve the original method name (`plot_histogram`) for backwards compatibility

File: rasa/nlu/test.py [hunk 3]
```diff
@@ -365,7 +365,7 @@ def plot_entity_confidences(
         if prediction not in (NO_ENTITY, target)
     ]
 
-    plot_utils.plot_histogram([pos_hist, neg_hist], title, hist_filename)
+    plot_utils.plot_comparative_histogram([pos_hist, neg_hist], title, hist_filename)
```
Comment (koernerfelicia) [inline rasa/nlu/test.py:368]: Please preserve the original method name (`plot_histogram`) for backwards compatibility

File: docs/docs/testing-your-assistant.mdx [hunk 4]
```diff
@@ -231,11 +232,17 @@ as well as providing an overall average. You can save these reports as JSON file
 
 The confusion matrix shows which intents are mistaken for others.
 Any samples which have been incorrectly predicted are logged and saved to a file called `errors.json` for easier debugging.
+<div align="center">
+    <img alt="image" src={useBaseUrl("/img/intent_confusion_matrix_example.png")} width="70%" />
+</div>
 
 The histogram allows you to visualize the confidence for all predictions,
 with the correct and incorrect predictions being displayed by blue and red bars respectively.
 Improving the quality of your training data will move the blue histogram bars to the right and the
```
Comment (koernerfelicia) [inline docs/docs/testing-your-assistant.mdx:241]: This is a little bit confusing now that the blue and red bars don't really move the right or left. Maybe something like:

"Improving the quality of your training data will move the blue histogram bars up the plot and the red histogram bars down the plot."

File: rasa/utils/plotting.py [hunk 5]
```diff
@@ -135,13 +135,51 @@ def plot_histogram(
...
-    plt.hist(hist_data, bins=bins, color=colors)
-    plt.xticks(bins)
-    plt.title(title)
-    plt.xlabel("Confidence")
-    plt.ylabel("Number of Samples")
-    plt.legend(["hits", "misses"])
+    binned_data_sets = [np.histogram(d, bins=bins)[0] for d in hist_data]
+
+    max_xlim = max(binned_data_sets[0] + binned_data_sets[1])
+    max_xlim += int(0.25 * max_xlim)  # padding
```
Comment (koernerfelicia) [inline rasa/utils/plotting.py:141]: Change from `int()` to `np.ceil()` to avoid 0 padding for values < 4

File: rasa/utils/plotting.py [hunk 6]
```diff
@@ -135,13 +135,51 @@ def plot_histogram(
...
-    plt.xlim([0, 1])
-    plt.hist(hist_data, bins=bins, color=colors)
-    plt.xticks(bins)
-    plt.title(title)
-    plt.xlabel("Confidence")
-    plt.ylabel("Number of Samples")
-    plt.legend(["hits", "misses"])
+    binned_data_sets = [np.histogram(d, bins=bins)[0] for d in hist_data]
+
+    max_xlim = max(binned_data_sets[0] + binned_data_sets[1])
```
Comment (koernerfelicia) [inline rasa/utils/plotting.py:140]: See my response -- I think we should have one max_xlim each for the left and right sides

File: rasa/utils/plotting.py [hunk 7]
```diff
@@ -135,13 +135,51 @@ def plot_histogram(
...
+        centers,
+        binned_data_sets[1],
+        height=heights,
+        align="center",
+        color=colors[1],
+        label="misses",
+    )
+    axes[1].set(title="Wrong")
+
+    axes[0].set(yticks=bins, xlim=(0, max_xlim))
```
Comment (koernerfelicia) [inline rasa/utils/plotting.py:166]: Add ylim = min non-empty bin here

**Comment by koernerfelicia** [comment]
intent_histogram
Hi @UTkzhang, @lymburn, @shadow-blade-X, @eanile: thank you for your work so far, I left a couple comments!
Two things regarding the x and y axes:

I tested this out on the [Sara test set](https://github.com/RasaHQ/rasa-demo/blob/master/tests/test_nlu.yml) and found that the scaling of the x axes makes it look like there are no errors (in reality, there are a few errors, and lots of correct, highly confident predictions). Even though I know and appreciate why you've given them the same limits, I think we should give the left and right sides their own axis limits.

Also, it'd be great if the y axis started at the minimum confidence bin, like so:
`min_ylim = bins[min((binned_data_sets[0] != 0).argmax(axis=0),
                        (binned_data_sets[1] != 0).argmax(axis=0))]`
intent_histogram

File: rasa/utils/plotting.py (hunk 1 above)
Comment (eanile) [reply rasa/utils/plotting.py:119]: Done

File: rasa/nlu/test.py (hunk 2 above)
Comment (eanile) [reply rasa/nlu/test.py:333]: Done

File: rasa/nlu/test.py (hunk 3 above)
Comment (eanile) [reply rasa/nlu/test.py:368]: Done

File: docs/docs/testing-your-assistant.mdx (hunk 4 above)
Comment (eanile) [reply docs/docs/testing-your-assistant.mdx:241]: Done

File: rasa/utils/plotting.py (hunk 5 above)
Comment (eanile) [reply rasa/utils/plotting.py:141]: Done

File: rasa/utils/plotting.py (hunk 6 above)
Comment (eanile) [reply rasa/utils/plotting.py:140]: Done.

File: rasa/utils/plotting.py (hunk 7 above)
Comment (eanile) [reply rasa/utils/plotting.py]: Done.

**Comment by eanile** [comment]
Hi @koernerfelicia, I've addressed all your comments, let me know if you think anything else should change.

**Comment by koernerfelicia** [comment]
@eanile I missed this on my first pass: but something's off about the centers. There should be a center between `0.95` and `1.0`, for example. In the examples I posted above the blue and red bars should really be between `0.95` and `1.0`. Maybe `.hist` with `orientation="horizontal"` is easier?

**Comment by UTkzhang** [comment]
@koernerfelicia it has been fixed :)

**Review by UTkzhang — APPROVED** [review]

**Comment by UTkzhang** [comment]
@koernerfelicia following up on the changes

File: docs/docs/testing-your-assistant.mdx [hunk 8]
```diff
@@ -231,11 +232,17 @@ as well as providing an overall average. You can save these reports as JSON file
...
+<div align="center">
+    <img alt="image" src={useBaseUrl("/img/intent_confusion_matrix_example.png")} width="70%" />
+</div>
 
 The histogram allows you to visualize the confidence for all predictions,
 with the correct and incorrect predictions being displayed by blue and red bars respectively.
-Improving the quality of your training data will move the blue histogram bars to the right and the
-red histogram bars to the left of the plot. It should also help in reducing the number red histogram bars itself.
+Improving the quality of your training data will move the blue histogram bars up the plot and the
+red histogram bars down the plot. It should also help in reducing the number red histogram bars itself.
```
Comment (koernerfelicia) [inline docs/docs/testing-your-assistant.mdx:242]:
```suggestion
red histogram bars down the plot. It should also help in reducing the number of red histogram bars itself.
```

**Review by koernerfelicia — COMMENTED** [review]
Looks great, thanks for addressing all my comments! There's a tiny typo in the docs (which was there before you touched them as well).

**Comment by UTkzhang** [comment]
@koernerfelicia Thanks!

**Comment by UTkzhang** [comment]
@koernerfelicia any ideas what this failure is about? https://github.com/RasaHQ/rasa/pull/7404/checks?check_run_id=1513018818

**Comment by koernerfelicia** [comment]
@UTkzhang, not sure, I just ran those same tests locally and everything seemed fine (it didn't even give me the warning you can see in the logs). I've restarted the jobs now, maybe it was just a fluke.

**Comment by UTkzhang** [comment]
Haha, thanks! Quoting Kevin from the office: "A fluke is one of the most common fishes in the ocean", really seems to apply to software engineering.

**Comment by UTkzhang** [comment]
@koernerfelicia hooray looks like it passed.

**Review by koernerfelicia — APPROVED** [review]

**Comment by koernerfelicia** [comment]
Haha! Thanks a lot for your work, we really appreciate it. Good luck with your class!
